### ISSUE TITLE
Works under windows?

### ISSUE THREAD

**Issue by Acelya-9028** [lewis6991/gitsigns.nvim#61]
**Describe the bug**

The signs column is not displayed.

**To Reproduce**

```vim
vim.cmd 'packadd paq-nvim'          -- Load Paq
local paq = require('paq-nvim').paq -- Import Paq

-- paq { 'junegunn/vim-easy-align' }   -- Provide mappings to align stuff
paq { 'lewis6991/gitsigns.nvim' }   --
paq { 'nvim-lua/plenary.nvim' }
-- paq { 'lifepillar/vim-gruvbox8' }   -- Add a beautiful theme
paq { 'savq/paq-nvim', opt=true }   -- Let Paq manage itself

require('gitsigns').setup {
  signs = {
    add          = {hl = 'DiffAdd'   , text = '│', numhl='GitSignsAddNr'},
    change       = {hl = 'DiffChange', text = '│', numhl='GitSignsChangeNr'},
    delete       = {hl = 'DiffDelete', text = '_', numhl='GitSignsDeleteNr'},
    topdelete    = {hl = 'DiffDelete', text = '‾', numhl='GitSignsDeleteNr'},
    changedelete = {hl = 'DiffChange', text = '~', numhl='GitSignsChangeNr'},
  },
  numhl = false,
  keymaps = {
    -- Default keymap options
    noremap = true,
    buffer = true,

    ['n ]c'] = { expr = true, "&diff ? ']c' : '<cmd>lua require\"gitsigns\".next_hunk()<CR>'"},
    ['n [c'] = { expr = true, "&diff ? '[c' : '<cmd>lua require\"gitsigns\".prev_hunk()<CR>'"},

    ['n <leader>hs'] = '<cmd>lua require"gitsigns".stage_hunk()<CR>',
    ['n <leader>hu'] = '<cmd>lua require"gitsigns".undo_stage_hunk()<CR>',
    ['n <leader>hr'] = '<cmd>lua require"gitsigns".reset_hunk()<CR>',
    ['n <leader>hp'] = '<cmd>lua require"gitsigns".preview_hunk()<CR>',
    ['n <leader>hb'] = '<cmd>lua require"gitsigns".blame_line()<CR>',

    -- Text objects
    ['o ih'] = ':<C-U>lua require"gitsigns".text_object()<CR>',
    ['x ih'] = ':<C-U>lua require"gitsigns".text_object()<CR>'
  },
  watch_index = {
    interval = 1000
  },
  sign_priority = 6,
  status_formatter = nil, -- Use default
}
```

No.

**Observed output**

No.

**Screenshots**

![nvim](https://user-images.githubusercontent.com/45818681/107981044-b0c4f080-6fc1-11eb-95de-a42808fd4fdd.png)

**Additional context**

```
Nom du système d’exploitation:              Microsoft Windows 10 Entreprise LTSC
Version du système:                         10.0.17763 N/A build 17763
```

**Comment by Acelya-9028** [comment]
Manually call `:lua require"gitsigns".preview_hunk()` display this:

![image](https://user-images.githubusercontent.com/45818681/107982172-d652f980-6fc3-11eb-98ee-5e228397503a.png)

**Comment by lewis6991** [comment]
The path handling currently assumes `/` as the path delimiter. For this to work on windows we'd need to change all the path handling to use the right delimiter. 

I don't have a Windows machine so I can't provide much support here. However I'm happy to help in any way I can.

**Comment by lewis6991** [comment]
Just opened #62. Could you give it a try?

**Comment by Acelya-9028** [comment]
Thank you very much. Sorry for the delay but I worked into this too and I found an another problem. More important I think.

I think the path problem can be fixed with a

```lua
local function get_buf_path(bufnr)
    return (uv.fs_realpath(api.nvim_buf_get_name(bufnr)) or
        api.nvim_buf_call(bufnr, function()
            return vim.fn.expand('%:p')
        end)):gsub("\\", "/")
end
```

Sorry for the Lua version but because Windows understand `/`, convert `\\` into `/` is good, I think.

But if you do this, Neovim crashes instantly 99% of the time. And I finally found where. The crash come from `os.tmpname()`, but not at any time, only after `local relpath, object_name, mode_bits, has_conflicts = await(git.file_info, file, toplevel)`.

Don't crash:

```lua
print(os.tmpname())
local relpath, object_name, mode_bits, has_conflicts = await(git.file_info, file, toplevel)
```

Crash:

```lua
local relpath, object_name, mode_bits, has_conflicts = await(git.file_info, file, toplevel)
print(os.tmpname())
```

I built an `init.lua` to reproduce the crash:

```lua
vim.cmd 'packadd paq-nvim'          -- Load Paq
local paq = require('paq-nvim').paq -- Import Paq

paq { 'nvim-lua/plenary.nvim' }

local Job = require('plenary/job')
local co = coroutine

function file_info(file, toplevel)
   return function(callback)
	  local relpath
	  local object_name
	  local mode_bits
	  local stage
	  local has_conflict = false
	  Job:new({
		 command = 'git',
		 args = {
			'--no-pager',
			'ls-files',
			'--stage',
			'--others',
			'--exclude-standard',
			file,
		 },
		 cwd = toplevel,
		 on_stdout = function(_, line)
			local parts = vim.split(line, '%s+')
			if #parts > 1 then
			   stage = tonumber(parts[3])
			   if stage <= 1 then
				  mode_bits = parts[1]
				  object_name = parts[2]
				  relpath = parts[4]
			   else
				  has_conflict = true
			   end
			else
			   relpath = parts[1]
			end
		 end,
		 on_exit = function(_, _)
			callback(relpath, object_name, mode_bits, has_conflict)
		 end,
	  }):start()
   end
end
	
function throttle_leading(ms, fn)
   local timer = vim.loop.new_timer()
   local running = false
   return function(...)
      if not running then
         timer:start(ms, 0, function()
            running = false
            timer:stop()
         end)
         running = true
         vim.schedule_wrap(fn)(...)
      end
   end
end

function async(func)
   return function(...)
      local params = { ... }
      return function(callback)
         local thread = co.create(func)
         local function step(...)
            local stat, ret = co.resume(thread, ...)
            assert(stat, ret)
            if co.status(thread) == "dead" then
               if callback then
                  callback(ret)
               end
            else
               (ret)(step)
            end
         end
         step(unpack(params))
      end
   end
end

function sync(func, ...)
    async(func)(...)()
end

throttle_leading(100, sync(function()
	local relpath, object_name, mode_bits, has_conflicts = co.yield(file_info('C:/Users/K2-18b/AppData/Local/nvim/init.lua', 'C:/Users/K2-18b/AppData/Local/nvim'))
	print('res:', relpath, object_name, mode_bits, has_conflicts)
	--print(os.tmpname())
end))
```

It's your code with only a missing `return function(...)` in `sync()` because `aync()` is never call with. I don't know why it's working in your code and not in the `init.lua` but this reproduce the wanted crash.

Thank you for your help.

**Comment by lewis6991** [comment]
Very interesting. My guess is that `os.tmpname()` doesn't work in callback threads on windows. Try adding `await_main()`just before to get back into the main thread.

**Comment by Acelya-9028** [comment]
Re sorry for the delay but yes you're right, it's good with a `await_main()` just before.

**Comment by lewis6991** [comment]
Good to here. So have you got the plugin to work on windows with all the mentioned fixes? If so a PR would be greatly appreciated!

**Comment by Acelya-9028** [comment]
I would like... But now `local staged_txt = await(git.get_staged_txt, toplevel, relpath, stage)` return `nil`. The temporary file is well created and valid but reading is broken. I thinks it's due to `table.insert(content, line)` but I have a lot of difficulty, I can't log anything because Neovim crash as soon as I add a print inside `on_stdout`. However I was able to read the first line with a `await_main()` but only the first.

I tried to execute this function synchronously but it's a fail. I have a lot of difficulty to understand how to do this.

**Comment by Acelya-9028** [comment]
Ok, I'm sorry for the previous comment. I delete it car very bad. I don't know how I did to log a `nil` but I can't reproduce it and a table is well there so I'm mystified.

There is just one thing, when the temporary file is written, `\r\n` is used insted of `\n` so the diff is broken but it's not a big problem. I don't know why because `writer:write(l .. '\n')` use `\n` but it's the case. Add `--ignore-space-at-eol` option to `git diff` fix this.

I make a PR.

**Comment by lewis6991** [comment]
https://github.com/lewis6991/gitsigns.nvim/pull/66

**Comment by Acelya-9028** [comment]
Thank you, I opened an another issue https://github.com/lewis6991/gitsigns.nvim/issues/69 for a related problem.
